core/connection.ts

```diff
@@ -243,67 +243,63 @@ export class Connection implements IASTNodeLocationWithBlock {
     return this.isConnected();
   }
 
-  /** Disconnect this connection. */
-  disconnect() {
-    const otherConnection = this.targetConnection;
-    if (!otherConnection) {
+  /**
+   * Disconnect this connection.
+   *
+   * @param setParent Whether to set the parent of the disconnected block or
+   *     not, defaults to true.
+   */
+  disconnect(setParent = true) {
+    const {parentConnection, childConnection} =
+        this.getParentAndChildConnections();
+    if (!parentConnection || !childConnection) {
       throw Error('Source connection not connected.');
     }
-    if (otherConnection.targetConnection !== this) {
-      throw Error('Target connection not connected to source connection.');
-    }
-    let parentBlock;
-    let childBlock;
-    let parentConnection;
-    if (this.isSuperior()) {
-      // Superior block.
-      parentBlock = this.sourceBlock_;
-      childBlock = otherConnection.getSourceBlock();
-      /* eslint-disable-next-line @typescript-eslint/no-this-alias */
-      parentConnection = this;
-    } else {
-      // Inferior block.
-      parentBlock = otherConnection.getSourceBlock();
-      childBlock = this.sourceBlock_;
-      parentConnection = otherConnection;
-    }
 
     const eventGroup = eventUtils.getGroup();
-    if (!eventGroup) {
-      eventUtils.setGroup(true);
-    }
-    this.disconnectInternal_(parentBlock, childBlock);
-    if (!childBlock.isShadow()) {
-      // If we were disconnecting a shadow, no need to spawn a new one.
-      parentConnection.respawnShadow_();
-    }
-    if (!eventGroup) {
-      eventUtils.setGroup(false);
-    }
-  }
+    if (!eventGroup) eventUtils.setGroup(true);
 
-  /**
-   * Disconnect two blocks that are connected by this connection.
-   *
-   * @param parentBlock The superior block.
-   * @param childBlock The inferior block.
-   */
-  protected disconnectInternal_(parentBlock: Block, childBlock: Block) {
     let event;
     if (eventUtils.isEnabled()) {
-      event =
-          new (eventUtils.get(eventUtils.BLOCK_MOVE))(childBlock) as BlockMove;
+      event = new (eventUtils.get(eventUtils.BLOCK_MOVE))(
+                  childConnection.getSourceBlock()) as BlockMove;
     }
     const otherConnection = this.targetConnection;
     if (otherConnection) {
       otherConnection.targetConnection = null;
     }
     this.targetConnection = null;
-    childBlock.setParent(null);
+    if (setParent) childConnection.getSourceBlock().setParent(null);
     if (event) {
       event.recordNew();
       eventUtils.fire(event);
     }
+
+    if (!childConnection.getSourceBlock().isShadow()) {
+      // If we were disconnecting a shadow, no need to spawn a new one.
+      parentConnection.respawnShadow_();
+    }
+
+    if (!eventGroup) eventUtils.setGroup(false);
+  }
+
+  /**
```

Comment:
Can do! Just wondering, do you want params and description to be generally required? Because we recently changed the lint config so we no longer require that with typescript.